vision/segmentation/er_ccl_clustering.go

```diff
@@ -0,0 +1,274 @@
+func (erCCL *ErCCLConfig) CheckValid() error {
+	if erCCL.MinPtsInPlane <= 0 {
+		return errors.Errorf("min_points_in_plane must be greater than 0, got %v", erCCL.MinPtsInPlane)
+	}
+	if erCCL.MinPtsInSegment <= 0 {
+		return errors.Errorf("min_points_in_segment must be greater than 0, got %v", erCCL.MinPtsInSegment)
+	}
+	if erCCL.MaxDistFromPlane == 0 {
+		erCCL.MaxDistFromPlane = 100
+	}
+	if erCCL.MaxDistFromPlane <= 0 {
+		return errors.Errorf("max_dist_from_plane must be greater than 0, got %v", erCCL.MaxDistFromPlane)
+	}
+	if erCCL.AngleTolerance > 180 || erCCL.AngleTolerance < 0 {
+		return errors.Errorf("max_angle_of_plane must between 0 & 180 (inclusive), got %v", erCCL.AngleTolerance)
+	}
+	if erCCL.ClusteringRadius < 0 {
+		return errors.Errorf("radius must be greater than 0, got %v", erCCL.ClusteringRadius)
+	}
+	if erCCL.Beta == 0 {
+		erCCL.Beta = 5
+	}
+	// going to have to add that the ground plane's normal vec has to be {0, 1, 0} or {0, 0, 1}
+	if !erCCL.NormalVec.IsUnit() {
+		return errors.Errorf("ground_plane_normal_vec should be a unit vector, got %v", erCCL.NormalVec)
+	}
+	if erCCL.NormalVec.Norm2() == 0 {
+		erCCL.NormalVec = r3.Vector{X: 0, Y: 0, Z: 1}
+	}
+	return nil
+}
+
+// ConvertAttributes changes the AttributeMap input into an ErCCLConfig.
+func (erCCL *ErCCLConfig) ConvertAttributes(am utils.AttributeMap) error {
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{TagName: "json", Result: erCCL})
+	if err != nil {
+		return err
+	}
+	err = decoder.Decode(am)
+	if err == nil {
+		err = erCCL.CheckValid()
+	}
+	return err
+}
+
+// NewERCCLClustering returns a Segmenter that removes the ground plane and returns a segmentation
+// of the objects in a point cloud using a connected components clustering algo described in the paper
+// "A Fast Spatial Clustering Method for Sparse LiDAR Point Clouds Using GPU Programming" by Tian et al. 2020.
+func NewERCCLClustering(params utils.AttributeMap) (Segmenter, error) {
+	// convert attributes to appropriate struct
+	if params == nil {
+		return nil, errors.New("config for ER-CCL segmentation cannot be nil")
+	}
+	cfg := &ErCCLConfig{}
+	err := cfg.ConvertAttributes(params)
+	if err != nil {
+		return nil, err
+	}
+	return cfg.ErCCLAlgorithm, nil
+}
+
+// ErCCLAlgorithm applies the connected components clustering algorithm directly on a given point cloud.
+func (erCCL *ErCCLConfig) ErCCLAlgorithm(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) {
+	// get next point cloud
+	cloud, err := src.NextPointCloud(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// run ransac, get pointcloud without ground plane
+	ps := NewPointCloudGroundPlaneSegmentation(cloud, erCCL.MaxDistFromPlane, erCCL.MinPtsInPlane, erCCL.AngleTolerance, erCCL.NormalVec)
+	// if there are found planes, remove them, and keep all the non-plane points
+	_, nonPlane, err := ps.FindGroundPlane(ctx)
+	if err != nil {
+		return nil, err
```

Comment:
You've done some nice code abstraction in your time here :) nice that doing this is now two lines!